File: ClaudeBuildPlan/src/lib/format.ts
import type { Database } from '@/api/database.types';

type MedicationFormulation = Database['public']['Enums']['medication_formulation'];

/**
 * Format a dose amount for display, picking the most natural unit
 * based on the medication's formulation.
 *
 *   liquid_suspension, infant_drops → "5 mL" (volume primary)
 *   chewable, oral_disintegrating   → "1 chewable" (count primary)
 */
export const formatDoseAmount = (input: {
  formulation: MedicationFormulation;
  amountMg?: number | null;
  amountVolumeMl?: number | null;
  unitCount?: number | null;
}): { primary: string; secondary?: string } => {
  const { formulation, amountMg, amountVolumeMl, unitCount } = input;

  if (formulation === 'liquid_suspension' || formulation === 'infant_drops') {
    if (amountVolumeMl != null) {
      return {
        primary: `${trimNumber(amountVolumeMl)} mL`,
        secondary: amountMg ? `${trimNumber(amountMg)} mg` : undefined,
      };
    }
  }

  if (formulation === 'chewable' || formulation === 'oral_disintegrating') {
    if (unitCount != null) {
      const noun = formulation === 'chewable' ? 'chewable' : 'tablet';
      return {
        primary: `${unitCount} ${unitCount === 1 ? noun : noun + 's'}`,
        secondary: amountMg ? `${trimNumber(amountMg)} mg` : undefined,
      };
    }
  }

  if (amountMg != null) {
    return { primary: `${trimNumber(amountMg)} mg` };
  }
  return { primary: '—' };
};

/**
 * "32.000" → "32"; "5.5" → "5.5"
 */
const trimNumber = (n: number): string => {
  const str = n.toFixed(3);
  return str.replace(/\.?0+$/, '');
};

/**
 * Compute initials from a display name, max 2 chars.
 *   "Ava M." → "AM"
 *   "Leonardo" → "LE"
 *   "Baby" → "BA"
 */
export const initialsFromName = (name: string): string => {
  if (!name) return '?';
  const parts = name.trim().split(/\s+/);
  if (parts.length === 1) return (parts[0]?.slice(0, 2) ?? '').toUpperCase();
  const first = parts[0]?.[0] ?? '';
  const last = parts[parts.length - 1]?.[0] ?? '';
  return (first + last).toUpperCase();
};

/**
 * Weight: prefer pounds in en-US, kilograms elsewhere.
 * For alpha we hardcode "lb" since design partners are US-based.
 *
 *   formatWeight(17_237) → "38 lb"
 */
export const formatWeightFromGrams = (
  grams: number | null | undefined,
  unit: 'lb' | 'kg' = 'lb',
): string => {
  if (grams == null) return '—';
  if (unit === 'kg') {
    const kg = grams / 1000;
    return `${kg.toFixed(1)} kg`;
  }
  const lb = grams / 453.592;
  return `${Math.round(lb)} lb`;
};

File: ios-native/Cappy/Domain/Format.swift
//
//  Format.swift
//  Cappy
//
//  Value formatting ported from app/src/lib/format.ts — dose amounts,
//  initials, and weight display.
//

import Foundation

enum CappyFormat {

    struct DoseAmount { let primary: String; let secondary: String? }

    /// Format a dose amount, picking the most natural unit for the formulation.
    static func doseAmount(formulation: MedicationFormulation,
                           amountMg: Double?,
                           amountVolumeMl: Double?,
                           unitCount: Int?) -> DoseAmount {
        switch formulation {
        case .liquidSuspension, .infantDrops:
            if let ml = amountVolumeMl {
                return DoseAmount(primary: "\(trim(ml)) mL",
                                  secondary: amountMg.map { "\(trim($0)) mg" })
            }
        case .chewable, .oralDisintegrating:
            if let count = unitCount {
                let noun = formulation == .chewable ? "chewable" : "tablet"
                let label = count == 1 ? noun : noun + "s"
                return DoseAmount(primary: "\(count) \(label)",
                                  secondary: amountMg.map { "\(trim($0)) mg" })
            }
        }
        if let mg = amountMg { return DoseAmount(primary: "\(trim(mg)) mg", secondary: nil) }
        return DoseAmount(primary: "—", secondary: nil)
    }

    /// "32.000" → "32"; "5.5" → "5.5".
    static func trim(_ n: Double) -> String {
        var s = String(format: "%.3f", n)
        while s.contains(".") && (s.hasSuffix("0") || s.hasSuffix(".")) {
            s.removeLast()
        }
        return s
    }

    /// Compute up-to-2-char initials from a display name. "Ava M." → "AM".
    static func initials(from name: String) -> String {
        let trimmed = name.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty { return "?" }
        let parts = trimmed.split(whereSeparator: { $0.isWhitespace })
        if parts.count == 1 {
            return String(parts[0].prefix(2)).uppercased()
        }
        let first = parts.first?.first.map(String.init) ?? ""
        let last = parts.last?.first.map(String.init) ?? ""
        return (first + last).uppercased()
    }

    /// Weight display. US design partners default to pounds.
    static func weight(fromGrams grams: Int?, unit: WeightUnit = .lb) -> String {
        guard let grams else { return "—" }
        switch unit {
        case .kg:
            return String(format: "%.1f kg", Double(grams) / 1000)
        case .lb:
            let lb = Double(grams) / 453.592
            return "\(Int(lb.rounded())) lb"
        }
    }

    enum WeightUnit { case lb, kg }
}
